### PR TITLE
Chapter 9

### PR DESCRIPTION
**Pull request by Moonforeva**

### CODE REVIEW

**Comment by Moonforeva** [comment]
Ready

File: app/assets/stylesheets/sessions.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/sessions.scss:3]: remove file that never used

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -3,16 +3,34 @@ def log_in user
...
 
   def logged_in?
     current_user.present?
   end
 
+  def forget user
+    user.forget
+    cookies.delete(:user_id)
+    cookies.delete(:remember_token)
+  end
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb:32]: remove ()

File: app/models/user.rb [hunk 3]
```diff
@@ -17,11 +18,28 @@ class User < ApplicationRecord
   class << User
     def digest string
       cost = ActiveModel::SecurePassword.min_cost ? Bcrypt::Engine::MIN_COST :
-                                                  Bcrypt::Engine.cost
-      Bcrypt::Password.create string, cost: cost
+                                                    Bcrypt::Engine.cost
+      Bcrypt::Password.create(string, cost: cost)
+    end
+    def new_token
+      SecureRandom.urlsafe_base64
```
Comment (saobv-2223) [inline app/models/user.rb:25]:
```suggestion

    def new_token
      SecureRandom.urlsafe_base64
```

File: app/models/user.rb [hunk 4]
```diff
@@ -17,11 +18,28 @@ class User < ApplicationRecord
...
-  
+ 
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
+  end
+
+  def authenticated?(remember_token)
+    return false if remember_digest.nil?
+    Bcrypt::Password.new(remember_digest).is_password?(remember_token)
```
Comment (saobv-2223) [inline app/models/user.rb:36]:
```suggestion
    return false if remember_digest.nil?
    
    Bcrypt::Password.new(remember_digest).is_password?(remember_token)
```

File: app/models/user.rb [hunk 5]
```diff
@@ -17,11 +18,28 @@ class User < ApplicationRecord
...
+    end
+    def new_token
+      SecureRandom.urlsafe_base64
     end
   end
-  
+ 
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
```
Comment (saobv-2223) [inline app/models/user.rb:31]: remove ()

File: app/models/user.rb [hunk 6]
```diff
@@ -17,11 +18,28 @@ class User < ApplicationRecord
...
+  end
+
+  def authenticated?(remember_token)
+    return false if remember_digest.nil?
+    Bcrypt::Password.new(remember_digest).is_password?(remember_token)
+  end
+
+  def forgets
+    update_attribute(:remember_digest, nil)
```
Comment (saobv-2223) [inline app/models/user.rb:40]: remove ()

File: config/routes.rb [hunk 7]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  get 'sessions/new'
```
Comment (saobv-2223) [inline config/routes.rb:2]: using "

File: package.json [hunk 8]
```diff
@@ -6,8 +6,18 @@
     "@rails/activestorage": "^6.0.0",
     "@rails/ujs": "^6.0.0",
     "@rails/webpacker": "4.3.0",
+<<<<<<< HEAD
+<<<<<<< HEAD
```
Comment (saobv-2223) [inline package.json:10]: resolve conflict

**Comment by Moonforeva** [comment]
Ready

File: app/assets/stylesheets/custom.scss [hunk 9]
```diff
@@ -177,3 +177,17 @@ input {
     color: $state-danger-text;
   }
 }
+
+.checkbox {
+  margin-top: -10px;
+  margin-bottom: 10px;
+  span {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss:184]:
```suggestion
  margin-bottom: 10px;

  span {
```

File: app/controllers/sessions_controller.rb [hunk 10]
```diff
@@ -5,6 +5,7 @@ def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate params[:session][:password]
       log_in user
+      params [:session][:remember_me] == "1" ? remember(user) : forget(user)
```
Comment (chungpv-1008) [inline app/controllers/sessions_controller.rb:8]: Create 'settings' variable for "1", then call back here

File: app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -3,16 +3,34 @@ def log_in user
...
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    if session[:user_id]
-      @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by(id: session[:user_id])
+    elsif user_id = cookies.signed[:user_id]
+      user = User.find_by(id: user_id)
```
Comment (chungpv-1008) [inline app/helpers/sessions_helper.rb:16]:
```suggestion
      user = User.find_by id: user_id
```

**Comment by Moonforeva** [comment]
Ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**